csrc/device_lower/pass/index.cpp

```diff
@@ -1742,10 +1743,240 @@ static MmaInputSmemSwizzle getSwizzleMode(TensorView* tv) {
       swizzle->inX()->extent()->evaluate().as<int64_t>() * 16);
 }
 
+// Get the ValGroup of the ID in consumer's loop domain that corresponds to the
+// innermost dimension in the allocation domain of tv. This ID must be
+// parallelized on Mma.
+ValGroup getInnerMmaLoopGroup(TensorView* tv, const MmaOp* mma) {
+  ValGraph& id_graph = GpuLower::current()->tensorIndexer().traversalGraph();
+  auto alloc_domain = id_graph.toGroups(
+      TensorDomain::noBroadcasts(tv->getMaybeAllocationDomain()));
+  auto loop_domain =
+      id_graph.toGroups(mma->out()->as<TensorView>()->getLoopDomain());
+
+  // Start from the innermost dim in the allocation domain, propagate all the
+  // way to the consumer's loop domain, and keep track of the inner dimension.
+  // After propagating, the inner dimension should be a dimension that is
+  // parallelized on Mma.
+  ValGroup inner = nullptr;
+  for (auto it = alloc_domain.rbegin(); it != alloc_domain.rend(); ++it) {
+    if (!(*it)->front()->as<IterDomain>()->isBroadcast()) {
+      inner = *it;
+      break;
+    }
+  }
+  NVF_ERROR(
+      inner != nullptr,
+      "Matmul with all broadcasting dimension is not supported yet.");
+
+  auto exprs =
+      ValGraphBFS::getExprsBetween(id_graph, loop_domain, alloc_domain);
+  while (!exprs.empty()) {
+    auto [expr, direction] = exprs.back();
```

Comment:
Why not `for (auto& [expr, direction] : exprs) {` instead?

Reply:
This is because we need to traverse `exprs` in the reverse order. Note that we can not just swap the `loop_domain` with the `alloc_domain` in the above `ValGraphBFS::getExprsBetween`, because `alloc_domain` does not fully cover loop domain, and if we do so, BFS will fail.

Reply:
> > because `alloc_domain` does not fully cover loop domain, and if we do so, BFS will fail
> 
> Ahhh OK. That makes sense. It will be nice to move to C++20 for cases like this so we can just do
> 
> ```c++
> for (auto& [expr, direction] : exprs | std::views::reverse) {
> ```

100% agree

Reply:
Why not just `ValGraphBFS::getExprsBetween(id_graph, alloc_domain, loop_domain);`?

Reply:
This does not work. Loop domain is bigger than the allocation domain and may contain IDs not in the allocation. This way, BFS will fail.